design/external-compaction.md

```diff
@@ -0,0 +1,249 @@
+
+On request for work from a compactor
+  1. For the queue requested, the coordinator gets the next tserver from the global queue summary map.
+  2. The Coordinator contacts the tablet server to reserve job
+  3. Return compaction job information to compactor
+  4. This must handle concurrent requests properly while minimizing the amount of locking.
+
+## Metrics
+
+TServers will each maintain their own internal queues for external compaction queues with all tablets.  If each tserver emits metrics about their local queues, then sum of all tserver metrics will give information about the external queues across the cluster.
+
+Compactor should emit metrics about the number of compactions running and which queue they are running on.
+
+## Scenarios
+
+Scenario 1
+
+ 1. Tablet T1 has files F1,F4,F5,F6
+ 1. The compaction of files F4,F5,F6 is queued on external executor EE2 for Tablet T1
+ 1. A new file F7 arrives for T1
+ 1. The compaction of F4,F5,F6 on EE2 is canceled and new a compaction of F4,F5,F6,F7 is queued on EE2 for Tablet T1
+ 1. A new file F8 arrives for T1
+ 1. The compaction of F4,F5,F6,F7 on EE2 is canceled and new a compaction of F4,F5,F6,F7,F8 is queued on EE2 for Tablet T1
+
+Scenario 2
+
+ 1. Tablet T1 has files F1,F2,F3,F4
+ 1. Compaction of files F1,F2,F3 is queued on internal executor IE1 for Tablet T1
+ 1. New files F5,F6,F7,F8,F9 arrive for tablet T1
+ 1. The compaction of F1,F2,F3 on IE1 is canceled and a new compaction of F1,F2,F3,F4,F5,F6,F7,F8,F9 is queued on external Executor EE2 for Tablet T1
+
+Scenario 3
+1. Tablet T1 has files F1,F2,F3
+2. Compaction of files F1,F1,F3 is queued on external executor EE2 for T1
+3. The table containing tablet T1 is deleted
+4. A compactor tries to compact F1,F2,F3 for tablet T1 and hopefully fails cleanly
+
+
+Scenario 4
+1. Compaction of files F1,F1,F3 is queued on external executor EE2 for T1
+2. The tablet server hosting tablet T1 dies
+3. Tablet T1 is loaded elsewhere 
+4. Compaction of files F1,F1,F3 is again queued on external executor EE2 for T1
+
+
+Scenario 5
+1. Compaction of files F1,F1,F3 is queued on external executor EE2 for T1
+2. A compactor starts compacting F1,F2,F3 for T1
+3. The tablet server hosting T1 dies
+4. The compactor compacting F1,F2,F3 for T1 finishes (T1 is not hosted anywhere yet)
+5. Another tablet server starts hosting T1
+6. The compactor is able to commit the compaction of F1,F2,F3 for T1
+
+
+Scenario 6
+1. Compaction of files F1,F1,F3 is queued on external executor EE2 for T1
+2. Compaction of files F4,F5,F6 is queued on external executor EE2 for T2
+3. A compactor start compacting F1,F2,F3 for T1
+4. The table hosting tablets T1 and T2 is taken offline
+5. The compactor working on F1,F2,F3 for T1 finishes and cleanly fails to commit the compaction because the table is offline
+6. The compaction for F4,F5,F6 is taken out of the queue, but cleanly fails to start because the table is offline.
```

Comment:
I know this design is strictly for online compactions but we could make offline compactions configurable. Maybe even have a separate queue for handling offline compactions.